ndb/src/google/cloud/ndb/model.py

```diff
@@ -1542,11 +1543,123 @@ def _get_for_dict(self, entity):
         return self._get_value(entity)
 
 
+def _validate_key(value, entity=None):
+    """Validate a key.
+
+    Args:
+        value (.Key): The key to be validated.
+        entity (Optional[Model]): The entity that the key is being validated
+            for.
+
+    Returns:
+        .Key: The passed in ``value``.
+
+    Raises:
+        .BadValueError: If ``value`` is not a :class:`.Key`.
+        KindError: If ``entity`` is specified, but the kind of the entity
+            doesn't match the kind of ``value``.
+    """
+    if not isinstance(value, Key):
+        raise exceptions.BadValueError("Expected Key, got {!r}".format(value))
+
+    if entity and type(entity) not in (Model, Expando):
+        if value.kind() != entity._get_kind():
+            raise KindError(
+                "Expected Key kind to be {}; received "
+                "{}".format(entity._get_kind(), value.kind())
+            )
+
+    return value
+
+
 class ModelKey(Property):
+    """Special property to store a special "key" for a :class:`Model`.
+
+    This is intended to be used as a psuedo-:class:`Property` on each
+    :class:`Model` subclass. It is **not** intended for other usage in
```

Comment:
I forgot to mention this originally. The **whole** purpose of this class is so that it can be used for `Model`:

```python
class Model:
    ...
    _key = ModelKey()
    ...
```

I wish we could make it non-public. WDYT? Should we have a list of "totally wish this wasn't public" members?

Reply:
I spent some time, not entirely successfully, trying to figure out what this was used for. [These snippets](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/6f5f3bcb81779679a24e0964a6c57c0c7deabfac/appengine/standard/ndb/entities/snippets.py#L98) show a `key` attribute of an entity being used directly in client code, so maybe it's not private? Or am I reading this incorrectly?

Reply:
Yes, I should've written

```python
class Model:
    ...
    _key = ModelKey()
    key = _key
    ...
```

however the public attribute `key` can be overridden. The whole point of the class attribute is so that queries like this can be created:

```python
key = ndb.Key(Book, 88)
query = Book.query(Book.key >= key)
```